File: Task 2/Java/logistics/src/main/java/com/logistics/Components/Product/Seeds.java
package com.logistics.Components.Product;

import java.util.Calendar;

public class Seeds implements Flora {
    protected float cost;
    protected String name;
    protected String supplier;
    protected float discountAmount;
    protected Packaging packaging;
    protected boolean isCultivated;
    private int expirationDays;

    public Seeds() {
        cost = 0;
        name = "";
        supplier = "";
        discountAmount = 1;
        packaging = new Packaging(0, 0);
        isCultivated = false;
        expirationDays = 0;
    }

    public Seeds(String name, float cost, String supplier, float discountAmount, int packingUnits, float costPerUnit, boolean isCultivated, int expirationDays) {
        this.name = name;
        this.cost = cost;
        this.supplier = supplier;
        this.discountAmount = discountAmount;
        this.packaging = new Packaging(packingUnits, costPerUnit);
        this.isCultivated = isCultivated;
        this.expirationDays = expirationDays;
    }

    // Copy Constructor
    public Seeds(Seeds other) {
        this.name = other.name;
        this.cost = other.cost;
        this.supplier = other.supplier;
        this.discountAmount = other.discountAmount;
        this.packaging = new Packaging(other.packaging);
        this.isCultivated = other.isCultivated;
        this.expirationDays = other.expirationDays;
    }

    // Clone method
    @Override
    public Seeds clone() {
        return new Seeds(this);
    }

    // Getters
    @Override public float getCost() { return cost; }
    @Override public String getName() { return name; }
    @Override public String getSupplier() { return supplier; }
    @Override public float getDiscountAmount(){return discountAmount;}
    @Override public float getPackagingCost() { return packaging.calculateTotalCost(); }
    @Override public boolean getIsCultivated() { return isCultivated; }
    public int getExpirationDays() { return expirationDays; }

    // Setters
    @Override public void setCost(float cost) { this.cost = cost; }
    @Override public void setName(String name) { this.name = name; }
    @Override public void setSupplier(String supplier) { this.supplier = supplier; }
    @Override public void setDiscountAmount(float discountAmount){this.discountAmount = discountAmount;}
    @Override public void setPackagingData(int packingUnits, float costPerUnit) { this.packaging = new Packaging(packingUnits, costPerUnit); }
    @Override public void setIsCultivated(boolean isCultivated) { this.isCultivated = isCultivated; }
    public void setExpirationDays(int expirationDays) { this.expirationDays = expirationDays; }

    @Override public float calculateDiscount() {
        // Get the current month
        int currentMonth = Calendar.getInstance().get(Calendar.MONTH);

        // Check if the current month is December 
        if (currentMonth == Calendar.DECEMBER && (discountAmount >= 0 && discountAmount <= 1)) {
            return (1 - discountAmount);
        } else {
            return 1;
        }
    }

    @Override public String display_product_info() {
        return  "\nName: " + name +
                "\nCost: " + cost +
                "\nSupplier: " + supplier +
                "\nDiscount Amount: " + calculateDiscount() +
                "\nPackaging Cost: " + getPackagingCost() +
                "\nCultivated: " + isCultivated +
                "\nExpiration Days: " + expirationDays;
    }
}

File: Task 3/C++Stock Viewer/Product/Seeds.cpp
#include "Seeds.h"
Seeds::Seeds() : Flora(){
    expirationDays = 0;
}

Seeds::Seeds(std::string name, float cost, std::string supplier, float discount_amount, int packing_units, float cost_per_unit, bool isCultivated, int expirationDays) 
        : Flora(name, cost, supplier, discount_amount, packing_units, cost_per_unit, isCultivated){
    this->expirationDays = expirationDays;
}

Seeds::Seeds(const Seeds& other) : Flora(other) {
    this->expirationDays = other.expirationDays;
}

std::string Seeds::display_product_info(){
    return Flora::display_product_info()+
            "\nExpiration Days: " + std::to_string(expirationDays);
}

// ========================================== PROTOBUFFERS
void Seeds::copyToProto(PBStockItem* message){
    PBSeeds* pbSeeds = message->mutable_seeds();
    pbSeeds->set_expirationdays(expirationDays);

    PBFlora* pbFlora = pbSeeds->mutable_flora();
    pbFlora->set_iscultivated(isCultivated);

    PBProduct* pbProduct = pbFlora->mutable_product();
    pbProduct->set_name(name);
    pbProduct->set_cost(cost);
    pbProduct->set_supplier(supplier);

    PBPackage* pbPackage = pbProduct->mutable_packaging();
    pbPackage->set_units(packaging.getUnits());
    pbPackage->set_costperunit(packaging.getCost_Per_Unit());
}

void Seeds::copyFromProto(const PBStockItem& message) {
    const PBSeeds& pbSeeds = message.seeds();
    name = pbSeeds.flora().product().name();
    cost = pbSeeds.flora().product().cost();
    supplier = pbSeeds.flora().product().supplier();
    packaging.setUnits(pbSeeds.flora().product().packaging().units());
    packaging.setCost_Per_Unit(pbSeeds.flora().product().packaging().costperunit());
    isCultivated = pbSeeds.flora().iscultivated();
    expirationDays = pbSeeds.expirationdays();
}